### DISCUSSION TITLE
ref var

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#854] [Language Ideas]
Hi C# team,
Ever since the introduction of "out var", I have found myself writing numerous methods that basically test a condition (actual return value) and return a value (out var) all in one step.  A constant limitation that I am running into with these methods though, is the ability to specify a default value.  I know that I can specify another parameter but that can make calling messy.  

Ideally, I would like to be able to do things like:
````
//TaskID & Total
var TaskCounts = new Dictionary<int, int>();
...
var TaskID = ...;
TaskCounts.TryGetValue(TaskID, ref var total = 0);

Console.WriteLine($"Task {TaskID} has a total of {total}");
````

I hope someone else finds the concept of ref var useful for when you want to "out var" something conditionally.

**Comment by svick** [comment]
I don't understand the example. If `TryGetValue` returns `false`, you don't use `Value`, so the default value does not matter.

And when you want to always have a value, you can use `??`:

```c#
long? personIndex = …;
var value = personIndex ?? 99;
Console.WriteLine(value);
```

Do you have an example that uses the default value and can't be replaced by `??`?

**Comment by TonyValenti** [comment]
I've updated my example.

**Comment by svick** [comment]
> I know that I can specify another parameter but that can make calling messy.

How so?

Compare:

```c#
TaskCounts.TryGetValue(TaskID, ref var total = 0);
TaskCounts.TryGetValue(TaskID, out var total, 0);
```

The second option doesn't look more messy to me.

(Also, if the default you want is `default`, then you can use code like `TaskCounts.TryGetValue(TaskID, out var total)` today.)

**Comment by jnm2** [comment]
Anyhow, I actually prefer

```c#
TaskCounts.TryGetValue(TaskID, out var total) ? total : 0
```
